samples/quantum-approximation-optimization/qaoa-for-quadratic-unconstrained-binary-optimization.ipynb

```diff
@@ -0,0 +1,801 @@
+   ],
+   "outputs": [],
+   "execution_count": null
+  },
+  {
+   "cell_type": "code",
+   "metadata": {},
+   "source": [
+    "import matplotlib.pyplot as plt\n",
+    "import numpy as np\n",
+    "\n",
+    "def plot_cost(cost, title):\n",
+    "    plt.figure(figsize=(12, 6))\n",
+    "    plt.plot(range(1, 11), np.array(cost), \"--o\")\n",
+    "    plt.xlabel(\"total time evolution (µs)\", fontsize=14)\n",
+    "    plt.ylabel(\"cost\", fontsize=14)\n",
+    "    plt.title(title, fontsize=16)\n",
+    "    plt.show()"
+   ],
+   "outputs": [],
+   "execution_count": null
+  },
+  {
+   "cell_type": "code",
+   "metadata": {},
+   "source": [
+    "plot_cost(cost_qutip_emulator, \"Qutip emulator\")"
+   ],
+   "outputs": [],
+   "execution_count": null
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": "## 4. Now let's run against Pasqal's emulator"
+  },
+  {
+   "cell_type": "code",
+   "metadata": {},
+   "source": [
+    "import json\n",
+    "\n",
+    "def prepare_input_data(seq):\n",
+    "    input_data = {}\n",
+    "    input_data[\"sequence_builder\"] = json.loads(seq.to_abstract_repr())\n",
+    "    return json.dumps(input_data)"
+   ],
+   "outputs": [],
+   "execution_count": null
+  },
+  {
+   "cell_type": "code",
+   "metadata": {},
+   "source": [
+    "def submit_job(target, seq):\n",
+    "    job = target.submit(\n",
+    "        input_data=prepare_input_data(seq), \n",
+    "        input_data_format=\"pasqal.pulser.v1\", \n",
+    "        output_data_format=\"pasqal.pulser-results.v1\",\n",
+    "        name=\"Pasqal sequence\",\n",
+    "        input_params={\n",
+    "            \"shots\": 10\n",
+    "        }\n",
+    "    )\n",
+    "\n",
+    "    print(f\"Queued job: {job.id}\")\n",
+    "    job.wait_until_completed()\n",
+    "    print(f\"Job completed with state: {job.details.status}\")\n",
+    "    result = job.get_results()\n",
+    "\n",
+    "    return result"
+   ],
+   "outputs": [],
+   "execution_count": null
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "Please set your `resource_id` and `location` below."
+  },
+  {
+   "cell_type": "code",
+   "metadata": {},
+   "source": [
+    "from azure.quantum import Workspace\n",
+    "workspace = Workspace(\n",
+    "            resource_id = \"\",\n",
+    "            location = \"\",\n",
+    ")\n",
+    "target = workspace.get_targets(name=\"pasqal.sim.emu-tn\")\n",
+    "\n",
+    "if isinstance(target, list):\n",
+    "    if len(target) == 0:\n",
+    "        raise RuntimeError(\"No emulators found for that name\")\n",
+    "    raise RuntimeError(target)\n",
+    "\n",
+    "# Open a session on the backend to group all jobs together\n",
+    "from azure.quantum.job.session import Session, SessionJobFailurePolicy \n",
+    "session = target.open_session(name=\"QUBO Problem\", job_failure_policy=SessionJobFailurePolicy.CONTINUE)\n",
```

Comment:
nit: could you please use 
```python
with target.open_session(name="QUBO Problem", job_failure_policy=SessionJobFailurePolicy.CONTINUE) as session
```
instead of explicit `session.close()`. In this case we can combine this cell with the one below and move the cell "Note: the next cell may take up to 5 minutes to complete." above this one.

Reply:
I have used the context manager and split the cells a bit differently, I hope you like it.